Enable DependencyTracker to evaluate interpolated paths

skipkayhil opened the pull request:
Previously, neither the PrismRenderParser nor the RipperRenderParser would consider an interpolated string as a dependency. The RubyTracker even included a line to explcitly filter out interpolated paths returned from the RipperRenderParser since they would end in a "/". However, the RubyTracker does include the ability to evaluate explicit "Template Dependency" comments with wildcard nodes.

This commit extends the RipperRenderParser and PrismRenderParser to convert interpolated strings into wildcard globs. Additionally, it changes the RubyTracker to evaluate wildcards the same for both implicit and explicit dependencies. This enables the RubyTracker to identify potential dependencies for interpolated renders, which it was previously unable to do.

cc @jhawthorn if you think this is a good idea I can try to apply it to ActionView::Precompiler as well

skipkayhil committed:
Enable RubyTracker to evaluate interpolated paths

Previously, neither the PrismRenderParser nor the RipperRenderParser
would consider an interpolated string as a dependency. The RubyTracker
even included a line to explcitly filter out interpolated paths returned
from the RipperRenderParser since they would end in a "/". However, the
RubyTracker does include the ability to evaluate explicit "Template
Dependency" comments with wildcard nodes.

This commit extends the RipperRenderParser and PrismRenderParser to
convert interpolated strings into wildcard globs. Additionally, it
changes the RubyTracker to evaluate wildcards the same for both implicit
and explicit dependencies. This enables the RubyTracker to identify
potential dependencies for interpolated renders, which it was previously
unable to do.

jhawthorn commented:
Seems like a good improvement! I am a little uncertain about us adding a feature to the RubyTracker that the ERBTracker does not have, as currently that's what is enabled by default (though I would love to switch over as we should not be parsing Ruby via regex). On the other hand it could be argued this is just another way `RubyTracker` is more accurate 😅.

skipkayhil commented:
> I am a little uncertain about us adding a feature to the RubyTracker that the ERBTracker does not have

I was afraid you would say this 🙈 haha

I'll see if I can figure out how hard it would be to add it to ERBTracker as well :+1:

skipkayhil committed:
Enable ERBTracker to evaluate interpolated paths

This ensures all three implementations (ERBTracker, RipperRenderParser,
and PrismRenderParser) are consistent in their ability to evaluate
interpolated paths.

skipkayhil committed:
Extract WildcardResolver to remove duplication

Since both the ERBTracker and RubyTracker now support resolving
interpolated template paths against the view_paths, the logic for this
resolution can be extracted to its own class.

skipkayhil commented:
@jhawthorn, I added the feature to the ERBTracker and updated the test to run against all three trackers. Anything else you think we should add?

jhawthorn commented on actionview/lib/action_view/dependency_tracker/wildcard_resolver.rb:
```diff
@@ -0,0 +1,32 @@
+        @view_paths = view_paths
+
+        @wildcard_dependencies, @explicit_dependencies =
+          dependencies.partition { |dependency| dependency.end_with?("/*") }
```
Does this mean we still only support the wildcards/interpolations when they're trailing? I think that's an okay limitation to have (maybe beneficial even to avoid matching too many/every template) but let's be sure to document that if so.

skipkayhil replied:
Yes, I just leveraged the existing code for explicit dependencies that let people do `Template Dependency: events/*`. I wanted to keep it simple for an initial implementation, and I definitely agree that wildcards in other positions could be a really easy performance footgun.

I added a new example to the `CacheHelper` docs to mention this new behavior, happy to make it more detailed if you think it would be beneficial!

jhawthorn committed:
Merge branch 'main' into hm-ruby-tracker-interpolation

jhawthorn committed:
Update CHANGELOG.md
